src/main/resources/include/hail/Region.h

```diff
@@ -0,0 +1,121 @@
+#ifndef HAIL_REGION_H
+#define HAIL_REGION_H 1
+
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <cstdlib>
+#include <cstring>
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+namespace hail {
+
+class Region;
+using RegionPtr = std::shared_ptr<Region>;
+
+class Region : public NativeObj {
+private:
+  static constexpr ssize_t kChunkCap = 64*1024;
+  static constexpr ssize_t kMaxSmall = 4*1024;
+  static constexpr ssize_t kNumBigToKeep = 4;
+  
+  struct CharArray {
+    char buf_[1];
+  };
+  
+  class UniqueChunk {
+   public:
+    std::unique_ptr<CharArray> ptr_;
```

Comment:
So you can have a unique_ptr<void> but you need a custom deleter.  I think we should have `unique_void_ptr` and `make_unique_void_ptr(ssize_t)` that use free/malloc.

https://stackoverflow.com/questions/39288891/why-is-shared-ptrvoid-legal-while-unique-ptrvoid-is-ill-formed/39288979#39288979

(I still don't quite understand the discussion there about shared_ptr and how it calls the destructor correctly but unique_ptr can't.)

Reply:
I reckon that destroying the CharArray is precisely going to call free().

But I already wrote it once the other way with explicit malloc/free and no unique_ptr,
so I can go back to that.